src/syntax/Stratification.jl

```diff
@@ -0,0 +1,295 @@
+        :($(shead...), $s => $t <= $a) => ([[DSLArgument(ss, t) for ss in shead] ; DSLArgument(s, t)], [DSLArgument(a, t)])#return (push!(Dict(unwrap_key_expression(ss, t) for ss in shead), unwrap_key_expression(s, t)), Dict(unwrap_key_expression(a, t)))
+
+        if mapping_pair.head == :tuple end => begin
+            middle_index = findfirst(x -> typeof(x) == Expr && length(x.args) == 3, mapping_pair.args) # still isn't specific enough
+            if isnothing(middle_index)
+                error("Malformed line $mapping_pair, could not find center.")
+            end
+            @match mapping_pair.args[middle_index] begin
+                :($stail => $t <= $ahead) => begin
+                    sdict = [[DSLArgument(ss, t) for ss in mapping_pair.args[1:middle_index-1]] ; DSLArgument(stail, t)]
+                    adict = [DSLArgument(ahead, t) ; [DSLArgument(as, t) for as in mapping_pair.args[middle_index+1:end]]]
+                    return (sdict, adict)
+                end
+                _ => "Unknown format found for match; middle three values formatted incorrectly."
+            end
+        end
+        _ => error("Unknown line format found in stratification notation.") 
+    end
+end
+
+
+
+
+function read_stratification_line_and_update_dictionaries!(line::Expr, strata_names::Dict{Symbol, Int}, type_names::Dict{Symbol, Int}, aggregate_names::Dict{Symbol, Int}, strata_mappings::Dict{Int, Int}, aggregate_mappings::Dict{Int, Int})
+    current_strata_symbol_dict, current_aggregate_symbol_dict = interpret_stratification_notation(line)
+
+    current_strata_dict = substitute_symbols(strata_names, type_names, current_strata_symbol_dict ; use_flags=true)
+    current_aggregate_dict = substitute_symbols(aggregate_names, type_names, current_aggregate_symbol_dict ; use_flags=true)
+    
+    if STRICT_MATCHES
+        @assert (all(x -> x ∉ keys(strata_mappings), keys(current_strata_dict))) "Attempt to overwrite a mapping in strata!"
+        # check that we're not overwriting a value which has already been assigned
+        merge!(strata_mappings, current_strata_dict) # accumulate dictionary keys
+
+
+        @assert (all(x -> x ∉ keys(aggregate_mappings), keys(current_aggregate_dict))) "Attempt to overwrite a mapping in aggregate!"
+        merge!(aggregate_mappings, current_aggregate_dict)
+
+    else 
+        mergewith!((x, y) -> x, strata_mappings, current_strata_dict) # alternatively, can use: only ∘ first
+        mergewith!((x, y) -> x, aggregate_mappings, current_aggregate_dict)
+    end
+
+end
+
+
+
+
+"""
+    @stratify (strata, type, aggregate) begin ... end
+
+    Ok, so the general idea here is:
+    1. Grab all names from strata, type and aggregate, and create dictionaries which map them to their indices
```

Comment:
I'm unsure how to separate it into functions without it having a ton of arguments.  Maybe I do some packing/unpacking for calls.

Reply:
Yeah maybe it'd be a bit ornerous given all of the params; it might be a pretty big refactoring task. Unless @Xiaoyan-Li thinks otherwise, maybe this can just stay as-is? It does have a lot of explanatory comments.

Reply:
I'd add I am liking the refactors I am seeing 😃

Reply:
In the last few commits I've factored out a few things by now operating vectors